core/src/agent/state.rs

```diff
@@ -30,6 +39,17 @@ pub struct AgentState {
     top_chain_header: Option<ChainHeader>,
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct AgentStateSnapshot {
+    top_chain_header: ChainHeader,
+}
+
+impl AgentStateSnapshot {
+    pub fn top_chain_header(&self) -> ChainHeader {
+        self.top_chain_header.clone()
```

Comment:
I would return a reference to the `top_chain_header` and leave it to the callee to clone it 🤔